File: app/lib/models/batch.dart
import 'customer.dart';
import 'product.dart';

enum BatchStatus { dangLam, chuaHoanThanh, hoanThanh }

BatchStatus batchStatusFromString(String value) {
  switch (value) {
    case 'chua_hoan_thanh':
      return BatchStatus.chuaHoanThanh;
    case 'hoan_thanh':
      return BatchStatus.hoanThanh;
    default:
      return BatchStatus.dangLam;
  }
}

String batchStatusLabel(BatchStatus status) {
  switch (status) {
    case BatchStatus.chuaHoanThanh:
      return 'Chưa hoàn thành';
    case BatchStatus.hoanThanh:
      return 'Hoàn thành';
    case BatchStatus.dangLam:
      return 'Đang làm';
  }
}

class Batch {
  final String id;
  final String code;
  final ProductRef product;
  final CustomerRef customer;
  final double? plannedQuantity;
  final BatchStatus status;
  final String? note;
  final double reportedQuantity;
  final double exportedQuantity;

  Batch({
    required this.id,
    required this.code,
    required this.product,
    required this.customer,
    this.plannedQuantity,
    required this.status,
    this.note,
    required this.reportedQuantity,
    this.exportedQuantity = 0,
  });

  factory Batch.fromJson(Map<String, dynamic> json) => Batch(
        id: json['_id'] as String,
        code: json['code'] as String,
        product: ProductRef.fromJson(json['product'] as Map<String, dynamic>),
        customer: CustomerRef.fromJson(json['customer'] as Map<String, dynamic>),
        plannedQuantity: (json['plannedQuantity'] as num?)?.toDouble(),
        status: batchStatusFromString(json['status'] as String? ?? 'dang_lam'),
        note: json['note'] as String?,
        reportedQuantity: (json['reportedQuantity'] as num?)?.toDouble() ?? 0,
        exportedQuantity: (json['exportedQuantity'] as num?)?.toDouble() ?? 0,
      );
}

File: backend/src/models/Batch.js
const mongoose = require("mongoose");

// LO_HANG: theo doi tien do va trang thai hoan thanh cua 1 lo san xuat
const batchSchema = new mongoose.Schema(
  {
    code: { type: String, required: true, trim: true, uppercase: true }, // so_lo, khop voi ProductionReport.batchNumber
    product: { type: mongoose.Schema.Types.ObjectId, ref: "Product", required: true },
    customer: { type: mongoose.Schema.Types.ObjectId, ref: "Customer", required: true },
    plannedQuantity: { type: Number, min: 0 }, // so luong ke hoach, khong bat buoc
    status: {
      type: String,
      enum: ["dang_lam", "chua_hoan_thanh", "hoan_thanh"],
      default: "dang_lam",
    },
    note: { type: String, trim: true },
    active: { type: Boolean, default: true },
    completedBy: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
    completedAt: { type: Date },
    createdBy: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
  },
  { timestamps: true }
);

batchSchema.index({ product: 1, customer: 1, code: 1 }, { unique: true });

module.exports = mongoose.model("Batch", batchSchema);
